kernel: system_work_q: Mark queue thread as essential

ExtremeGTX opened the pull request:
Fixes #70293 
Marking sysworkq as essential, so when it fails, the system will halt instead of continuously working, and dependent components stay in a broken state.

ExtremeGTX commented:
Question to reviewers: is it better to make a flag part of the `k_work_queue_config` struct ?
```
struct k_work_queue_config {
	const char *name;
	bool no_yield;
	bool essential;
};
```

so it can be passed to `k_work_queue_start` ?

cfriedt reviewed (commented):
Ideally flags could be set before the workqueue thread starts, so the proposed config change makes sense to me.

cfriedt commented on kernel/system_work_q.c:
```diff
@@ -26,10 +26,12 @@ static int k_sys_work_q_init(void)
 	};
 
+	k_sys_work_q.thread.base.user_options |= K_ESSENTIAL;
 	k_work_queue_start(&k_sys_work_q,
```
Your suggestion of adding it to the config probably makes the most sense - I should have clarified earlier

TaiJuWu reviewed (approved):
Make sense for me.

dcpleung commented on kernel/system_work_q.c:
```diff
@@ -26,10 +26,12 @@ static int k_sys_work_q_init(void)
 		.no_yield = IS_ENABLED(CONFIG_SYSTEM_WORKQUEUE_NO_YIELD),
 	};
 
+	k_sys_work_q.thread.base.user_options |= K_ESSENTIAL;
```
This would not work as `k_work_queue_start()` calls `k_thread_create()` which eventually calls `z_init_thread_base()` which overrides `user_options`.

ExtremeGTX committed:
kernel: work_queue: make thread essential flag configurable

Allow the creator of a work_queue instance to choose whether
the work_queue thread should be marked as ESSENTIAL or not.

Signed-off-by: Mohamed ElShahawi <ExtremeGTX@hotmail.com>

peter-mitsis commented on kernel/work.c:
```diff
@@ -761,6 +761,10 @@ void k_work_queue_start(struct k_work_q *queue,
 		k_thread_name_set(&queue->thread, cfg->name);
 	}
 
+	if (cfg->essential) {
+		queue->thread.base.user_options |= K_ESSENTIAL;
+	}
```
Perhaps there is a reason that I am missing, but why not conditionally pass K_ESSENTIAL as the _options_ parameter in the call to _k_thread_create()_ above?  That is something like, ...

`    k_thread_create(&queue->thread, ...., prio, cfg->essential ? K_ESSENTIAL : 0, K_FOREVER);`

That would eliminate some messing around with k_thread internals.

ExtremeGTX replied:
since it is acceptable to pass `cfg=NULL` this can cause a crash, i modified the condition to also check if `cfg != NULL`

nashif commented on include/zephyr/kernel.h:
```diff
@@ -4002,6 +4002,11 @@ struct k_work_queue_config {
+	/** Control whether the work queue thread should be marked as
+	 * essential thread.
+	 */
+	bool essential;
```
instead of making `essential` a global config for all work queues, it makes more sense to have a 'system' config, indicating this is a system work queue, then based on this, you can set the essential flag on system work queues.

ExtremeGTX replied:
> instead of making essential a global config for all work queues, it makes more sense to have a 'system' config, indicating this is a system work queue, then based on this, you can set the essential flag on system work queues.

you mean to have something like:
```c
/* new type in addition to `k_work_queue_config` */
struct k_sys_work_queue_config {
	const char *name;
	bool no_yield;
	bool essential;
};

void k_sys_work_queue_start(struct k_work_q *sys_queue,
			k_thread_stack_t *stack,
			size_t stack_size,
			int prio,
			const struct k_sys_work_queue_config *cfg)
{
	struct k_work_queue_config cfg = {
		.name = cfg->name,
		.no_yield = cfg->no_yield
	};
	k_work_queue_start(sys_queue, stack, stack_size, prio)
	sys_queue.thread.base.user_options |= K_ESSENTIAL;
}
```



But I would like to say that the idea behind including `essential` in the queue config is to allow the user to mark their own queue also as essential not only system work queue. for example we have a system where it uses a custom Bluetooth transport which has its own workq and if this workq thread crashed it doesn't make sense for our system to continue to work. WDYT ?

nashif replied:
no, I mean

```
/* new type in addition to `k_work_queue_config` */
struct k_sys_work_queue_config {
	const char *name;
	bool no_yield;
	bool system_wq;
};
```

and later on: 

```
 k_thread_create(&queue->thread, ...., prio, cfg->system_wq ? K_ESSENTIAL : 0, K_FOREVER);
```

> But I would like to say that the idea behind including `essential` in the queue config is to allow the user to mark their own queue also as essential not only system work queue.

that is not what the title/description of the PR says...

cfriedt replied:
Isn't there only 1 system work queue? Essential might be the better term 🤷🏼‍♂️

ExtremeGTX replied:
> and later on:
 `k_thread_create(&queue->thread, ...., prio, cfg->system_wq ? K_ESSENTIAL : 0, K_FOREVER);`

That means `bool system_wq` will be added to `k_work_queue_config` type without creating the new type `k_**sys**_work_queue_config`, otherwise we will need to have 2 implementations of `k_work_queue_start()`

nashif replied:
> Isn't there only 1 system work queue? Essential might be the better term 🤷🏼‍♂️

there is only one, and hence the suggestion by only reading the PR description and the related issue, however, there is another commit in this PR that is not mentioned in the PR description which allows essential to be set on any work queue. I was commenting based on the PR description.

andyross reviewed (dismissed):
Yeah, this seems way too heavyweight.  ESSENTIAL threads just abort the system when they terminate, that's all they do.  Essnetially (heh) the feature is a klugey and weak form of watchdogging.  I don't think we want to be getting fancy with it, or really pushing it very hard as a good tunable for app design.  Quite honestly I wouldn't cry if it was just removed.

That said, I think most of the proposals here are reasonable.  For sure apps will have their own work queues that they will want to be able to terminate or shut down.  So the tunable needs to be per-queue.  Whether we call that flag "essential" or "system" seems like a taste thing.  Again I'm not a big fan of the API at all, so I won't express a preference.

+1 to all the suggested variants, basically.  Feel free to assume-refreshed-approval if it gets pushed again.

ExtremeGTX committed:
kernel: system_work_q: Mark queue thread as essential

Marking sysworkq as essential, so when it fails, the system will halt
instead of continuously working, and dependent components stay
in a broken state.

Signed-off-by: Mohamed ElShahawi <ExtremeGTX@hotmail.com>

andyross reviewed (approved):
Refresh +1, though as mentioned I have no strong feelings here and think _THREAD_ESSENTIAL is a sort of weak feature we shouldn't be pushing.

ExtremeGTX commented:
@andyross could you please share what can work as a better alternative to the `essential` feature ?
